dev/tests/static/testsuite/Magento/Test/Integrity/Dependency/DependencyProvider.php

```diff
@@ -0,0 +1,207 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Magento\Test\Integrity\Dependency;
+
+use Magento\Framework\App\Utility\Files;
+use Magento\Framework\Component\ComponentRegistrar;
+
+abstract class DependencyProvider
+{
+    /**
+     * Types of dependency between modules.
+     */
+    const TYPE_HARD = 'hard';
+
+    /**
+     * The identifier of dependency for mapping.
+     */
+    const MAP_TYPE_DECLARED = 'declared';
+
+    /**
+     * The identifier of dependency for mapping.
+     */
+    const MAP_TYPE_FOUND = 'found';
+
+    /**
+     * @var array
+     */
+    protected $mapDependencies = [];
+
+    /**
+     * @var array
+     */
+    protected $packageModuleMapping = [];
+
+    /**
+     * Retrieve array of dependency items.
+     *
+     * @param $module
+     * @param $type
+     * @param $mapType
+     * @return array
+     */
+    abstract protected function getDeclaredDependencies(string $module, string $type, string $mapType);
```

Comment:
this is the template pattern, but we try to stay away from abstract and protected
Use composition instead